modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Output.scala

```diff
@@ -12,30 +12,89 @@
  */
 package com.snowplowanalytics.snowplow.enrich.fs2
 
+import java.time.Instant
+import java.nio.charset.StandardCharsets.UTF_8
+
 import cats.effect.Concurrent
 import fs2.Pipe
 
-import com.snowplowanalytics.snowplow.badrows.BadRow
 import com.snowplowanalytics.snowplow.enrich.common.outputs.EnrichedEvent
+import com.snowplowanalytics.snowplow.badrows.{BadRow, Failure, Payload => BadRowPayload}
+import com.snowplowanalytics.snowplow.enrich.common.utils.ConversionUtils
 
-/** Represents the three different outputs of enrichment that should be delivered to three different sinks */
-sealed trait Output
+/** An item that can be sunk into any one of the enrichment sinks: Bad, Good, or Pii */
+sealed trait Output[+Bad, +Good, +Pii]
 
 object Output {
 
-  case class Bad(badRow: BadRow) extends Output
-  case class Good(event: EnrichedEvent) extends Output
-  case class Pii(event: EnrichedEvent) extends Output
+  /** An item destined for the Bad Sink */
+  case class Bad[B](badRow: B) extends Output[B, Nothing, Nothing]
+
+  /** An item destined for the Good Sink */
+  case class Good[G](event: G) extends Output[Nothing, G, Nothing]
+
+  /** An item destined for the Pii Sink */
+  case class Pii[P](event: P) extends Output[Nothing, Nothing, P]
 
-  /** A variant of fs2's observeEither, sinking three output types into three Pipe's */
-  def sink[F[_]: Concurrent](
-    bad: Pipe[F, BadRow, Unit],
-    good: Pipe[F, EnrichedEvent, Unit],
-    pii: Pipe[F, EnrichedEvent, Unit]
-  ): Pipe[F, Output, Unit] =
+  type Parsed = Output[BadRow, EnrichedEvent, EnrichedEvent]
+  type Serialized = Output[Array[Byte], Array[Byte], Array[Byte]]
+
+  /** Sends a bad/good/pii item into the appropriate sink */
+  def sink[F[_]: Concurrent, B, G, P](
+    bad: Pipe[F, B, Unit],
+    good: Pipe[F, G, Unit],
+    pii: Pipe[F, P, Unit]
+  ): Pipe[F, Output[B, G, P], Unit] =
     _.observe(_.collect { case Bad(br) => br }.through(bad))
       .observe(_.collect { case Good(event) => event }.through(good))
       .observe(_.collect { case Pii(event) => event }.through(pii))
       .drain
 
+  /**
+   * Serializes events and bad rows into byte arrays. Good/Pii events exceeding
+   *  a maximum size after serialization are converted to bad events.
+   */
+  val serialize: Parsed => Serialized =
+    serializeGood.andThen(serializePii).andThen(serializeBad)
+
+  private def serializeGood[P]: Output[BadRow, EnrichedEvent, P] => Output[BadRow, Array[Byte], P] = {
+    case Good(g) =>
+      serializeEnriched(g).fold(Bad(_), Good(_))
+    case b: Bad[BadRow] => b
```

Comment:
I think mentioning a generic type in pattern match might be a bit confusing for people not knowing about type erasure.

Reply:
Will change to

```scala
case b @ Bad(_) => b
case p @ Pii(_) => p
```